decisions/2025-02-package-upgrade-solution.md

```diff
@@ -0,0 +1,136 @@
+# Package Upgrade Solutions
+
+<!-- This template is inspired by
+https://github.com/GoogleCloudPlatform/emblem/tree/main/docs/decisions -->
+
+* **Status:** Draft <!--/ Approved / Rejected / Superseded-->
+* **Last Updated:** 2025-02-04
+* **Objective:** Solve issues with dependency compatibility.
+
+## Context & Problem Statement
+
+The Flutter SDK pins packages instead of using the caret constraint to allow for
+minor upgrades.
+
+## Goals
+
+To determine a viable package version resolution strategy that ensures all of
+our dependencies remain up to date and compatible.
+
+### Non-goals
+
+Although this should be a goal, based on the state of the ecosystem, we should
+not necessarily ensure that our upgrade strategy prevents breakage, but rather
+test for when it does. There have consistently been breaking changes across
+minor version bumps in the ecosystem, and if we want to keep packages up to date,
+ensuring we receive security fixes when they come, we will have to accept the
+reality of breaking changes occurring when the semver increase dictates otherwise.
+
+## Other considerations <!-- optional -->
+
+### Flutter SDK Pinned Packages
+
+In the Dart SDK docs they mention the reason for package pins in the Flutter SDK
+[here](https://github.com/dart-lang/sdk/blob/2ac59922d2e6ab5f62ceb8725ffad96cfd147c68/docs/Flutter-Pinned-Packages.md?plain=1#L13-L15).
+This statement indicates that teams maintaining Flutter & Dart have no intention
+of respecting semver, as they don't even trust themselves to respect it.
+More elaboration on the consequences of this in the following options.
+
+### at_mono resolution issues
+
+at_mono depends on all major library packages in our ecosystem to ensure
+compatibility, and lately the Flutter packages have been incompatible with the
+Dart ones. The reason for this is that dependabot has bumped meta to `^1.16.0`
+yet the latest stable flutter SDK has `flutter_test` pinned to `1.15.0`. This
+is an ongoing issue that occurs as we upgrade dependencies across the ecosystem.
+
+### Package tree version resolution tooling
+
+Another thing to note is the differences between tooling options for maintaining
+dependencies:
+
+- [melos](https://melos.invertase.dev/)
```

Comment:
Linter is whining about using `-` rather than `*`

Reply:
Once we've reached a final decision, I'll update the document and fix all lint issues.